libbpf-tools/schedsnoop.bpf.c

```diff
@@ -0,0 +1,241 @@
+// SPDX-Licence-Identifier: GPL-2.0
+// Copyright (c) 2020 Alibaba Cloud
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "schedsnoop.h"
+
+#define TASK_RUNNING 0x0000
+#define TASK_REPORT_MAX 0x0100
+
+const volatile pid_t targ_pid = 0;
+const volatile int trace_syscall = 0;
+bool targ_exit = false;
+int trace_on = -1;
+
+struct{
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+struct{
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, NR_ENTRY_MAX);
+	__type(key, struct si_key);
+	__type(value, u64);
+} syscall_info_maps SEC(".maps");
+
+static __always_inline void set_trace_on(int cpu)
+{
+	trace_on = cpu;
+}
+
+static __always_inline void set_trace_off(void)
+{
+	trace_on = -1;
+}
+
+static __always_inline int should_trace(int cpu)
+{
+	return trace_on == cpu;
+}
+
+static __always_inline void emit_trace(void *ctx, struct trace_info* ti)
+{
+	ti->ts = bpf_ktime_get_ns();
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, ti, sizeof(*ti));
+}
+
+static __always_inline void emit_and_update(void *ctx, struct trace_info* ti,
+	       	struct si_key* sik)
+{
+	__u64 *last_ts;
+	__u64 ts = bpf_ktime_get_ns();
+	switch (ti->type) {
+	case TYPE_SYSCALL_ENTER:
+		bpf_map_update_elem(&syscall_info_maps, sik, &ts, BPF_ANY);
+		break;
+	case TYPE_SYSCALL_EXIT:
+		last_ts = bpf_map_lookup_elem(&syscall_info_maps, sik);
+		if(!last_ts)
+			return;

```

Comment:
when a syscall exits we need to know when does it start and calculate the duration. If we couldn't find any record in syscall_info_maps, we just drop this event instead of emit it to perfbuffer. That's why we return here.

Reply:
ah, right, makes sense